File: src/main/py/server/workflow.py
# ------------------------------------------------------------------------------
# Workflow objects managed by the YADAGE Server
# ------------------------------------------------------------------------------

# ------------------------------------------------------------------------------
# Workflow descriptor
#
# Workflow descriptors maintain basic information about workflow instances
# managed by the YADAGE server. This information includes the unique workflow
# identifier, the descritpive (user-provided) workflow name, and the current
# state of the workflow. More comprehensive workflow classes will inherit from
# this base class.
#
# While the state of a workflow can be derived from the workflow Json object,
# storing it separately is intended to speed up the generation of workflow
# listing containing each workflow's state in the Web API. It is expected that
# the engine using the instance manager ensures that the state that is stored
# with the instance is identical to the state that would be derived from the
# Json object.
# ------------------------------------------------------------------------------
class WorkflowDescriptor(object):
    # --------------------------------------------------------------------------
    # Initialize the identfifier, name, and state.
    #
    # identifier::string
    # name::string
    # state::string
    # --------------------------------------------------------------------------
    def __init__(self, identifier, name, state):
        self.identifier = identifier
        self.name = name
        self.state = state


# ------------------------------------------------------------------------------
# Workflow Database Instance
#
# Workflow instances that as managed by a workflow instance manager. Extends
# the workflow descriptor with the YADAGE Json representation of the workflow.
# ------------------------------------------------------------------------------
class WorkflowDBInstance(WorkflowDescriptor):
    # --------------------------------------------------------------------------
    # Initialize the identfifier, name, state, and Json object containing the
    # YADAGE workflow representation.
    #
    # identifier::string
    # name::string
    # state::string
    # workflow_json::JsonObject (for YADAGE workflow)
    # --------------------------------------------------------------------------
    def __init__(self, identifier, name, state, workflow_json):
        super(WorkflowDBInstance, self).__init__(identifier, name, state)
        self.workflow_json = workflow_json


# ------------------------------------------------------------------------------
# Full workflow instance object. Extends the workflow descriptor with the
# workflow graph, list of rules, list of applied rules, and list of applicable
# rule identifier.#
# ------------------------------------------------------------------------------
class WorkflowInstance(WorkflowDescriptor):
    # --------------------------------------------------------------------------
    # Initialize the identfifier, name, state, dag, rules, applied rules and
    # applicable rule identifier. At this stage all ADAGE objects are simply
    # Json objects.
    #
    # identifier::string
    # name::string
    # state::string
    # dag::JsonObject
    # rules::JsonObject
    # applied_rules::JsonObject
    # applicable_rules::[string]
    # submittable_nodes::[strig]
    # --------------------------------------------------------------------------
    def __init__(self, identifier, name, state, dag, rules, applied_rules, applicable_rules, submittable_nodes):
        super(WorkflowInstance, self).__init__(identifier, name, state)
        self.dag = dag
        self.rules = rules
        self.applied_rules = applied_rules
        self.applicable_rules = applicable_rules
        self.submittable_nodes = submittable_nodes


# ------------------------------------------------------------------------------
# CLASS: WorkflowState
#
# Representation of a workflow state. At this point the state is simple
# represented by a unique descriptive name.
# ------------------------------------------------------------------------------
class WorkflowState(object):
    # --------------------------------------------------------------------------
    # Initialize the name of a workflow state.
    #
    # name: string
    # --------------------------------------------------------------------------
    def __init__(self, name):
        self.name = name

    # --------------------------------------------------------------------------
    # Unambiguous printable representation of the workflow state
    # --------------------------------------------------------------------------
    def __repr__(self):
        return '<WorkflowState: {}>'.format(self.name)

    # --------------------------------------------------------------------------
    # Readable printable representation of the workflow state
    # --------------------------------------------------------------------------
    def __str__(self):
        return self.name


# ------------------------------------------------------------------------------
# Define possible workflow states. A workflow can be in either of four
# different states: RUNNING (i.e., submitted task that is running), WAITING
# (i.e., waiting for user interaction), FAILED (i.e., execution of at least on
# task falied), and SUCCESS (i.e., workflow successfully completed).
# ------------------------------------------------------------------------------
WORKFLOW_RUNNING = WorkflowState('RUNNING')
WORKFLOW_WAITING = WorkflowState('WAITING')
WORKFLOW_FAILED  = WorkflowState('FAILED')
WORKFLOW_SUCCESS = WorkflowState('SUCCESS')


# ------------------------------------------------------------------------------
# Dictionary of workflow states
# ------------------------------------------------------------------------------
WORKFLOW_STATES = {
    'WAITING': WORKFLOW_WAITING,
    'RUNNING': WORKFLOW_RUNNING,
    'FAILED':  WORKFLOW_FAILED,
    'SUCCESS': WORKFLOW_SUCCESS,
}

File: static/js/yadage/workflow.js
/**
 * Workkflow state codes
 */
var STATE_RUNNING = 'RUNNING';
var STATE_WAITING = 'WAITING';
var STATE_FAILED = 'FAILED';
var STATE_FINISHED = 'SUCCESS';


/**
 * ADAGE Workflow descriptor. Wraps a descriptor object returned by the ADAGE
 * Server API.
 */
var WorkflowDescriptor = function(descriptor_obj) {
    /**
     * Unique workflow identifier.
     *
     * @type: string
     */
    this.id = descriptor_obj.id;
    /**
     * User-provided workflow name.
     *
     * @type: string
     */
    this.name = descriptor_obj.name;
    /**
     * Current state of the workflow. Possible values are given by the workflow
     * state codes.
     *
     * @type: string
     */
    this.state = descriptor_obj.state;
    /**
     * Workflow URL. Extracted from the list of references.
     *
     * @type: string
     */
    this.url = getSelfReference(descriptor_obj.links);
};


/**
 * ADAGE Workflow - Contains methods and properties for an ADAGE workflow
 * object returned by the ADAGE Server API.
 */
var Workflow = function(workflow_obj) {
    /**
     * Directed acyclic graph of defined, running, completed, and failed
     * workflow tasks.
     *
     * @type: ADAGEDAG
     */
    this.dag = new ADAGEDAG(workflow_obj.dag);
    /**
     * Unique workflow identifier.
     *
     * @type: string
     */
    this.id = workflow_obj.id;
    /**
     * User-provided workflow name.
     *
     * @type: string
     */
    this.name = workflow_obj.name;
    /**
     * Current state of the workflow. Possible values are given by the workflow
     * state codes.
     *
     * @type: string
     */
    this.state = workflow_obj.state;
    /**
     * Workflow URL. Extracted from the list of references.
     *
     * @type: string
     */
    this.url = getSelfReference(workflow_obj.links);
    /**
     * Dictionary of rule objects
     */
    this.rules = {};
    for (var i_rule = 0; i_rule < workflow_obj.rules.length; i_rule++) {
        var rule = workflow_obj.rules[i_rule];
        this.rules[rule.id] = rule.rule;
    }
    /**
     * List of applicable rules
     *
     * @type: [string]
     */
    this.applicable_rules = workflow_obj.applicableRules;
    /**
     * List of applicable nodes
     *
     * @type: [string]
     */
    this.submittable_nodes = [];
    for (var i_node = 0; i_node < workflow_obj.submittableNodes.length; i_node++) {
        var node = this.dag.getNode(workflow_obj.submittableNodes[i_node]);
        this.submittable_nodes.push(node);
    }
};


/**
 * Listing of workflow descriptors. Splits a list of descriptors returned by the
 * ADAGE Server API based on workflow state.
 */
var WorkflowListing = function(wf_list_obj) {
    /**
     * Dictionary containing workflows grouped by state.
     *
     * @type: {[WorkflowDescriptor]}
     */
    var states = [STATE_RUNNING, STATE_WAITING, STATE_FAILED, STATE_FINISHED];
    this.categories = {};
    for (var i_state = 0; i_state < states.length; i_state++) {
        this.categories[states[i_state]] = [];
    }
    /**
     * Dictionary containing workflow descriptors indexed by their identifier.
     *
     * @type: {WorkflowDescriptor}
     */
    this.lookup = {};
    /**
     * List of workflow descriptors
     *
     * @type: [WorkflowDescriptor]
     */
    for (var i_wf = 0; i_wf < wf_list_obj.length; i_wf++) {
        var wf = new WorkflowDescriptor(wf_list_obj[i_wf]);
        this.categories[wf.state].push(wf);
        this.lookup[wf.id] = wf;
    }
    // Sort elements in category lists by id
    for (var i_state = 0; i_state < states.length; i_state++) {
        this.categories[states[i_state]].sort(function(wf1, wf2) {
            return wf1.id.localeCompare(wf2.id);
        });
    }
};
